sbi/utils/conditional_density.py

```diff
@@ -321,3 +327,317 @@ def _normalize_probs(probs: Tensor, limits: Tensor) -> Tensor:
+        """
+
+        # extract and rescale means, mixture componenets and covariances
+        nn = self.net
+        dist = nn._distribution
+
+        if context == None:
+            encoded_x = nn._embedding_net(self.default_x)
+        else:
+            encoded_x = nn._embedding_net(torch.tensor(context, dtype=torch.float32))
+
+        logits, m, prec, sumlogdiag, _ = dist.get_mixture_components(encoded_x)
+        norm_logits = logits - torch.logsumexp(logits, dim=-1, keepdim=True)
+
+        scale = nn._transform._transforms[0]._scale
+        shift = nn._transform._transforms[0]._shift
+
+        means_transformed = ((m - shift) / scale).detach()
+        
+        A = scale * torch.eye(means_transformed.shape[2])
+        precision_factors_transformed = torch.cholesky(A@prec@A)
+        
+        self.logits = norm_logits.detach()
+        self.means = means_transformed.detach()
+        self.precs = precision_factors_transformed.detach()
+        self.sumlogdiag = torch.sum(torch.log(torch.diagonal(self.precs, dim1=2, dim2=3)),dim=2).detach()
+        
+        return norm_logits, means_transformed, precision_factors_transformed, sumlogdiag
+
+    @staticmethod
+    def mulnormpdf(X: Tensor, mu: Tensor, cov: Tensor) -> Tensor:
+        """Evaluates the PDF for the multivariate Guassian distribution.
+
+        Args:
+            X: torch.tensor with inputs/entries row-wise. Can also be a 1-d array if only a
+                single point is evaluated.
+            mu: torch.tensor with center/mean, 1d array.
+            cov: 2d torch.tensor with covariance matrix.
+
+        Returns:
+            prob: Probabilities for entries in `X`.
+        """
+
+        # Evaluate pdf at points or point:
+        if X.ndim == 1:
+            X = torch.atleast_2d(X)
+        sigma = torch.atleast_2d(cov)  # So we also can use it for 1-d distributions
+
+        N = mu.shape[0]
+        ex1 = torch.inverse(sigma) @ (X - mu).T
+        ex = -0.5 * (X - mu).T * ex1
+        if ex.ndim == 2:
+            ex = torch.sum(ex, axis=0)
+        K = 1 / torch.sqrt(
+            torch.pow(2 * torch.tensor(3.14159265), N) * torch.det(sigma)
+        )
+        return K * torch.exp(ex)
+
+    def log_prob(self, X: Tensor, individual=False) -> Tensor:
+        """Evaluates the Mixture of Gaussian (MoG)
+        probability density function at a value x.
+
+        Args:
+            X: Values at which to evaluate the MoG pdf.
+            individual: If True the probability density is returned for each cluster component.
+
+        Returns:
+            log_prob: Log probabilities at values specified by X.
+        """
+        prec = self.precs@self.precs.transpose(3,2)
+
+        self.net.eval() # leakage correction requires eval mode
+        log_factor = torch.log(self.leakage_correction(x=self.default_x))
+        
+        log_prob = mdn.log_prob_mog(X,self.logits, self.means, prec, self.sumlogdiag) # only works for single samples
```

Comment:
I noticed, that `log_prob_mog` does not seem to work with batches of samples.

Reply:
Are you sure? It does not work if you follow the shapes outlined [here](https://github.com/mackelab/pyknos/blob/b8631df59a4f4053097264589c3b0b797a985016/pyknos/mdn/mdn.py#L181-L188)?

Reply:
fixed it. Wasn't aware the parameters had to be repeated for each batch example.